app/models/custom_field/calculated_value.rb

```diff
@@ -73,29 +75,75 @@ def formula_string
+        .where.not(id: id)
+        .visible
+    end
+
+    def validate_referenced_custom_fields_allowed
+      # We can only validate used custom fields from a high-level perspective, since at this point in validation,
+      # we do not have a project context to check against. So we cannot check if the custom fields are actually
+      # enabled for a project and visible to a non-admin user.
+      formula_cfs = CustomField.where(id: cf_ids_used_in_formula(formula_string)).pluck(:id)
+      allowed_cfs = usable_custom_field_references_for_formula.pluck(:id)
+
+      surplus_cfs = formula_cfs - allowed_cfs
+
+      if surplus_cfs.any?
+        custom_field_names = CustomField.where(id: surplus_cfs).pluck(:name)
+        errors.add(:formula, :invalid_custom_fields, custom_fields: custom_field_names.join(", "))
+        return
+      end
+
+      true
+    end
+
     private
 
+    def calculator
+      @calculator ||= Dentaku::Calculator.new(case_sensitive: true)
+    end
+
+    def valid_formula_syntax?
+      # Attempt to parse the formula. If no error is returned, the formula is syntactically valid.
+      calculator.ast(formula_str_without_patterns)
+      true
+    rescue Dentaku::ParseError, Dentaku::TokenizerError
+      false
+    end
+
     def formula_contains_only_allowed_characters?
       # List of allowed characters in a formula. This only performs a very basic validation.
-      # Allowed characters are:
-      # + - / * ( ) whitespace digits and decimal points
-      # Additionally, the formula may contain references to custom fields in the form of `cf_123` where 123 is the ID of
-      # the custom field.
+      # The allowed characters are:
+      # Our mathematical operators, whitespace, digits and decimal points
+      # Additionally, the formula may contain references to custom fields in the form of `{{cf_123}}`
+      # where 123 is the ID of the custom field.
       # Once this basic validation passes, the formula will be parsed and validated by Dentaku, which builds an AST
       # and ensures that the formula is really valid. A welcome side effect of the basic validation done here is that
       # it prevents built-in functions from being used in the formula, which we do not want to allow.
-      allowed_chars = %w[+ - / * ( )] + [" "]
-      allowed_tokens = /\A(cf_\d+|\d+\.?\d*|\.\d+)\z/
+      allowed_chars = OPERATORS + [" "]
+      allowed_tokens = /\A(\{\{cf_\d+}}|\d+\.?\d*|\.\d+)\z/
 
       formula_string.split(Regexp.union(allowed_chars)).reject(&:empty?).all? do |token|
         token.match?(allowed_tokens)
       end
     end
 
     # Returns a list of custom field IDs used in the formula.
-    # For a formula like `2 + cf_12 + cf_4` it returns `[12, 4]`.
+    # For a formula like `2 + {{cf_12}} + {{cf_4}}` it returns `[12, 4]`.
     def cf_ids_used_in_formula(formula_str)
-      formula_str.scan(/\bcf_(\d+)\b/).flatten.map(&:to_i)
+      formula_str.scan(/\{\{cf_(\d+)}}/).flatten.compact.map(&:to_i)
```

Comment:
I don't think compact is needed here, also `flatten` can be removed by using lookaheads:
```suggestion
      formula_str.scan(/(?<=\{\{cf_)\d+(?=}})/).map(&:to_i)
```

Reply:
Oh, that's great! TIL thank you.